api/src/auth/RoomAuth.ts

```diff
@@ -2,18 +2,32 @@ import { sign, verify } from 'jsonwebtoken';
 import Room from '../core/Room';
 import { roomTokenSecret } from '../Environment';
 import { randomUUID } from 'crypto';
+import { RoomAction } from '../types/RoomAction';
 
 export type RoomTokenPayload = {
     roomSlug: string;
     uuid: string;
+    isSpectating: boolean;
+    isMonitor: boolean;
 };
 
+type Permissions = Partial<{
+    [P in keyof RoomTokenPayload as RoomTokenPayload[P] extends boolean
+        ? P
+        : never]: RoomTokenPayload[P];
+}>;
```

Comment:
I got a stroke, reading this. This should either be simplified or you should add a comment that details what the types does